book/src/dev/rfcs/0010-v5-transaction.md

```diff
@@ -0,0 +1,224 @@
+
+V4 and V5 transactions both support sapling data but they are implemented differently. By this reason we need to split sapling data types into V4 and V5. We do this by using `enum` data types.
+
+The order of some of the fields changed from V4 to V5. For example the `lock_time` and `expiry_height` were moved above the transparent inputs and outputs.
+
+All of the changes proposed in this document are only to the `zebra-chain` crate.
+
+To highlight changes all document comments from the code snippets in the [reference section](#reference-level-explanation) were removed.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+We start by looking how a V4 (already implemented) transaction is represented in Zebra.
+
+This transaction version is specified by the protocol in the first table of [Transaction Encoding and Consensus](https://zips.z.cash/protocol/nu5.pdf#txnencodingandconsensus).
+
+
+```
+V4 {
+    inputs: Vec<transparent::Input>,
+    outputs: Vec<transparent::Output>,
+    lock_time: LockTime,
+    expiry_height: block::Height,
+    value_balance: Amount,
+    joinsplit_data: Option<JoinSplitData<Groth16Proof>>,
+    shielded_data: Option<ShieldedData>,
+}
+```
+
+Currently the `ShieldedData` type is defined in `zebra-chain/src/transaction/shielded_data.rs` as follows:
+
+```
+pub struct ShieldedData {
+    pub first: Either<Spend, Output>,
+    pub rest_spends: Vec<Spend>,
+    pub rest_outputs: Vec<Output>,
+    pub binding_sig: Signature<Binding>,
+}
+```
+
+We know by protocol (2nd table of [Transaction Encoding and Consensus](https://zips.z.cash/protocol/nu5.pdf#txnencodingandconsensus)) that V5 transactions will support sapling data however we also know by protocol that spends ([Spend Description Encoding and Consensus](https://zips.z.cash/protocol/nu5.pdf#spendencodingandconsensus), See †) and outputs ([Output Description Encoding and Consensus](https://zips.z.cash/protocol/nu5.pdf#outputencodingandconsensus), See †) fields change from V4 to V5.
+
+Here we have the proposed changes for V4 transactions:
+
+```
+V4 {
+    ...
+    sapling_value_balance: Amount,
+    ...
+    sapling_shielded_data: Option<SaplingShieldedData::V4>,
+}
+```
+
+In order to mantain the order of the fields that we will need for Serialization/Deserialization we keep the `value_balance` in the transaction for V4 but we rename it to `sapling_value_balance` for clarity.
+
+`ShieldedData` is currently defined and implemented in `src/transaction/shielded_data.rs`. We propose to rename this file to `src/transaction/sapling_shielded_data.rs`. Inside we will change `ShieldedData` into a `SaplingShieldedData` enum with `V4` and `V5` variants.
+
+```
+pub enum SaplingShieldedData {
+    /// Either a spend or output description.
+    ///
+    /// Storing this separately ensures that it is impossible to construct
+    /// an invalid `SaplingShieldedData` with no spends or outputs.
+    /// ...
```

Comment:
If this is now Sapling-specific, should we move it into the `sapling` module, like the `Spend`s and `Output`s? That would allow us to remove the `Sapling` name from the type. @teor2345

Reply:
Yes, that sounds good.